### PR TITLE
Add Event to Pallet Transaction Payment

### LINKED ISSUE #11595: Add `TransactionPayment` event to Pallet Tranascation Payment
There have been many issues from various people, all asking how to get the transaction fee of a Substrate transaction.

Currently, the only events that are emitted are those from the Balances pallet, which are not specific enough to determine the actual transaction fee in all cases.

We should add a simple event to the Transaction Payment Pallet which emits exactly the final transaction fee, tip, and user to clarify all of this.

Something like:

```rust
#[pallet::event]
#[pallet::generate_deposit(pub(super) fn deposit_event)]
pub enum Event<T: Config> {
	TransactionPayment { account: T::AccountId, fee: BalanceOf<T>, tip: BalanceOf<T> },
}
```

### PR DESCRIPTION
**Pull request by hzy1919**
Fixes #11595 .

And fix tests in Pallet Balance/Executive/Asset-tx-payment .

polkadot companion: https://github.com/paritytech/polkadot/pull/5677
cumulus companion: https://github.com/paritytech/cumulus/pull/1367

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
Should need a companion to add the event to the Polkadot runtimes?

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -158,6 +158,7 @@ mod tests;
 mod benchmarking;
 mod tests_composite;
 mod tests_local;
+#[cfg(test)]
```
Comment (shawntabrizi) [inline frame/balances/src/lib.rs:161]: is this needed?

**Review by shawntabrizi — APPROVED** [review]
Looks good! Are you able to make a polkadot and cumulus companion?

File: frame/balances/src/lib.rs (hunk 1 above)
Comment (hzy1919) [reply frame/balances/src/lib.rs:161]: yeah, the compiler issued that warning.

**Comment by hzy1919** [comment]
> 
@shawntabrizi  Ok, I will try to make a polkadot and cumulus companion.

File: frame/transaction-payment/src/lib.rs [hunk 2]
```diff
@@ -734,6 +744,7 @@ where
 			T::OnChargeTransaction::correct_and_deposit_fee(
 				&who, info, post_info, actual_fee, tip, imbalance,
 			)?;
+			Pallet::<T>::deposit_event(Event::<T>::TransactionPaid { who, fee: actual_fee, tip });
```
Comment (joepetrowski) [inline frame/transaction-payment/src/lib.rs:747]: `actual_fee` includes the tip here, so I would make it explicit in the event/documentation.
https://github.com/paritytech/substrate/blob/ef661427da2f9cec68f871ddf8e2de32bc0402ac/frame/transaction-payment/src/types.rs#L81

File: frame/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -321,6 +324,13 @@ pub mod pallet {
 		}
 	}
 
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A transaction has been paid.
+		TransactionPaid { who: T::AccountId, fee: BalanceOf<T>, tip: BalanceOf<T> },
```
Comment (joepetrowski) [inline frame/transaction-payment/src/lib.rs:331]:
```suggestion
		/// A transaction fee `actual_fee`, of which `tip` was added to the minimum inclusion fee, has been paid by `who`.
		TransactionFeePaid { who: T::AccountId, actual_fee: BalanceOf<T>, tip: BalanceOf<T> },
```

File: frame/transaction-payment/src/lib.rs (hunk 2 above)
Comment (joepetrowski) [reply frame/transaction-payment/src/lib.rs:747]: I.e. the inclusion fee is `actual_fee - tip`

**Comment by shawntabrizi** [comment]
bot merge

**Comment by hzy1919** [comment]
> 

@shawntabrizi  OK, thanks for your help. I have added my polkadot address .

**Pull request merged**